jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ExtUtils.java

```diff
@@ -35,10 +65,186 @@ public static String toUri(JarEntryFile jarEntryFile) {
         }
     }
 
+    public static boolean isJarResourceUri(URI uri) {
+        return uri != null && JDT_SCHEME.equals(uri.getScheme()) && CONTENTS_AUTHORITY.equals(uri.getAuthority());
+    }
+
+    public static IJarEntryResource createJarResource(URI uri) throws JavaModelException {
+        String handleId = uri.getQuery();
+        IPackageFragmentRoot packageRoot = (IPackageFragmentRoot) JavaCore.create(handleId);
+        String path = uri.getPath();
+
+        // if the file exists in the java packages
+        String[] segments = StringUtils.split(path, "/");
+        String packageName = StringUtils.join(Arrays.asList(segments).subList(0, segments.length - 1), '.');
+        IPackageFragment packageFragment = packageRoot.getPackageFragment(packageName);
+        if (packageFragment != null && packageFragment.exists()) {
+            Object[] objs = packageFragment.getNonJavaResources();
+            for (Object obj : objs) {
+                if (obj instanceof IJarEntryResource) {
+                    IJarEntryResource child = (IJarEntryResource) obj;
+                    if (child instanceof JarEntryFile && child.getFullPath().toPortableString().equals(path)) {
+                        return child;
+                    }
+                }
+
+            }
+        }
+        Object[] resources = packageRoot.getNonJavaResources();
+
+        for (Object resource : resources) {
+            if (resource instanceof JarEntryFile) {
+                JarEntryFile file = (JarEntryFile) resource;
+                if (file.getFullPath().toPortableString().equals(path)) {
+                    return file;
+                }
+            }
+            if (resource instanceof JarEntryDirectory) {
+                JarEntryDirectory directory = (JarEntryDirectory) resource;
+                return findFileInJar(directory, path);
+            }
+        }
+        return null;
+    }
+
+    public static JarEntryFile findFileInJar(JarEntryDirectory directory, String path) {
+        for (IJarEntryResource child : directory.getChildren()) {
+            if (child instanceof JarEntryFile && child.getFullPath().toPortableString().equals(path)) {
+                return (JarEntryFile) child;
+            }
+            if (child instanceof JarEntryDirectory) {
+                JarEntryFile file = findFileInJar((JarEntryDirectory) child, path);
+                if (file != null) {
+                    return file;
+                }
+            }
+        }
+        return null;
+    }
+
     public static IPath removeProjectSegment(String projectElementName, IPath path) {
         if (projectElementName.equals(path.segment(0))) {
             return path.removeFirstSegments(1).makeRelative();
         }
         return path;
     }
+
+    public static PackageNode createNodeForProject(IJavaElement javaElement) {
```

Comment:
Why all these functions are in the utility class? The file will be bloated with static methods. Please organize the logic clearly.